### PR TITLE
[AAP-73139] Removes RUNTIME_FEATURE_FLAGS_UI

### PR DESCRIPTION
**Pull request by huffmanca**
## Description
We've been informed that RUNTIME_FEATURE_FLAGS_UI isn't used, so removing it from the repository.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Test update
- [ ] Refactoring (no functional changes)
- [ ] Development environment change
- [x] Configuration change

## Self-Review Checklist

- [x] I have performed a self-review of my code
- [x] I have added relevant comments to complex code sections
- [x] I have updated documentation where needed
- [x] I have considered the security impact of these changes
- [x] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [x] I have tested the changes in my local environment

### Steps to Test
1. Start the Gateway instance.
2. Navigate to $INSTANCE/api/gateway/v1/settings/feature_flags/
3. Confirm that `RUNTIME_FEATURE_FLAGS_UI` is no longer present.

### Expected Results

`RUNTIME_FEATURE_FLAGS_UI` should not be present in the settings/feature_flags/ section.

### Required Actions

- [ ] Requires documentation updates
  
- [ ] Requires downstream repository changes
  
- [ ] Requires infrastructure/deployment changes
  
- [ ] Requires coordination with other teams
  
- [ ] Blocked by PR/MR: #XXX
  

## Summary by CodeRabbit

* **Removals**
  * The runtime feature-flags UI preference was removed and is no longer available in system preferences.

* **Data Migration**
  * A migration deletes any stored instances of that removed preference; rollbacks will not recreate it.

* **Tests**
  * Tests updated to stop asserting the removed UI preference.

* **CI / Workflows**
  * Added an auto-close PR workflow and a manual backport workflow trigger.

### CODE REVIEW

File: aap_gateway_api/registered_preferences.py [hunk 1]
```diff
@@ -503,22 +503,6 @@
...
-    preference_type="bool",
-    help_text=_(
-        "Controls whether the feature flags UI page is displayed. This is currently set through settings. "
-        "If you'd like to change this please update the RUNTIME_FEATURE_FLAGS_UI setting in your configuration."
-    ),
-    label=_('Feature Flags UI Page'),
-    read_only=True,
-    encrypted=False,
-    settings_bound=True,
-)
```
Comment (john-westcott-iv) [inline aap_gateway_api/registered_preferences.py:520]: I think you need a migration if you want to actually strip this out of the DB.

File: aap_gateway_api/tests/feature_flags/test_feature_flag_api.py [hunk 2]
```diff
@@ -453,7 +453,6 @@ def test_feature_flags_detail(admin_api_client, feature_flag, value):
     'preference',
     [
         ("RUNTIME_FEATURE_FLAGS"),
```
Comment (john-westcott-iv) [inline aap_gateway_api/tests/feature_flags/test_feature_flag_api.py:455]: With the `register()` call removed and only one preference remaining in the parametrize list, this ends up as:

```python
@pytest.mark.parametrize(
    "preference",
    [
        ("RUNTIME_FEATURE_FLAGS"),
    ],
)
```

Pre-existing style nit: `("RUNTIME_FEATURE_FLAGS")` is just parenthesized string, not a single-element tuple (would need a trailing comma inside: `("RUNTIME_FEATURE_FLAGS",)`). Works fine here since parametrize handles both strings and tuples, but worth noting.

**Review by john-westcott-iv — COMMENTED** [review]
## 📋 Code Review Summary

Clean removal of the unused `RUNTIME_FEATURE_FLAGS_UI` setting from defaults, preference registration, test fixture, and test parametrization. All 4 files are consistently cleaned up.

**Files Reviewed**: 4 files
**Comments Posted**: 2 review comments (1 existing + 1 new)

### 🔍 Issues Found
- 0 security/safety issues
- 1 correctness question (need a migration to clean up DB row?)
- 1 pre-existing style nit (parametrize tuple syntax)

### Overall Assessment
**Needs clarification** — The code removal itself is correct and complete. The open question is whether a migration is needed to remove the orphaned preference row from the database. Without it, stale data will persist in the `preferences` table for existing deployments.

### General Feedback
- All 4 touchpoints (defaults, preferences, conftest, tests) are consistently cleaned up
- No remaining references to `RUNTIME_FEATURE_FLAGS_UI` appear to be left behind in the changed files

**Comment by john-westcott-iv** [comment]
# DVT Report: [AAP-73139] Removes RUNTIME_FEATURE_FLAGS_UI

**PR**: [https://github.com/ansible/jewel/pull/5](https://github.com/ansible/jewel/pull/5)
**JIRA**: [AAP-73139](https://redhat.atlassian.net/browse/AAP-73139)
**Date**: 2026-04-27T20:20:00Z
**Gateway Branch**: dvt/jewel-pr-5
**DAB Branch**: 

---

## Summary

| Metric | Value |
|--------|-------|
| Total Tests | 8 |
| Passed | 8 |
| Failed | 0 |
| Pass Rate | 100% |

---

## PR Description Summary

Removes the unused RUNTIME_FEATURE_FLAGS_UI setting from defaults, preference registration, test fixture, and test parametrization.

---

## JIRA Acceptance Criteria Coverage

**Source**: [AAP-73139](https://redhat.atlassian.net/browse/AAP-73139)

_No acceptance criteria found on the JIRA ticket._

---


## Test Results

### 1. RUNTIME_FEATURE_FLAGS_UI removed from preferences API :white_check_mark: (PASS)
- **Category**: happy_path
- **Description**: Verify that RUNTIME_FEATURE_FLAGS_UI no longer appears in the feature_flags settings section
- **Expected Status**: 200
- **Actual Status**: 200

<details>
<summary>curl command</summary>

```bash
curl -sk -u admin:admin https://localhost:8000/api/gateway/v1/settings/feature_flags/
```
</details>

### 2. RUNTIME_FEATURE_FLAGS still present in preferences :white_check_mark: (PASS)
- **Category**: regression
- **Description**: Verify that the sibling RUNTIME_FEATURE_FLAGS preference was not accidentally removed
- **Expected Status**: 200
- **Actual Status**: 200

<details>
<summary>curl command</summary>

```bash
curl -sk -u admin:admin https://localhost:8000/api/gateway/v1/settings/feature_flags/
```
</details>

### 3. Feature flags list endpoint works :white_check_mark: (PASS)
- **Category**: regression
- **Description**: Verify the feature_flags list API still returns flags correctly after the removal
- **Expected Status**: 200
- **Actual Status**: 200

<details>
<summary>curl command</summary>

```bash
curl -sk -u admin:admin https://localhost:8000/api/gateway/v1/feature_flags/
```
</details>

### 4. Feature flag detail endpoint works :white_check_mark: (PASS)
- **Category**: regression
- **Description**: Verify a specific feature flag can still be retrieved by ID
- **Expected Status**: 200
- **Actual Status**: 200

<details>
<summary>curl command</summary>

```bash
curl -sk -u admin:admin https://localhost:8000/api/gateway/v1/feature_flags/1/
```
</details>

### 5. Settings sections list includes feature_flags :white_check_mark: (PASS)
- **Category**: regression
- **Description**: Verify the feature_flags section still appears in the settings index
- **Expected Status**: 200
- **Actual Status**: 200

<details>
<summary>curl command</summary>

```bash
curl -sk -u admin:admin https://localhost:8000/api/gateway/v1/settings/
```
</details>

### 6. Unauthenticated access to preferences returns 401 :white_check_mark: (PASS)
- **Category**: negative
- **Description**: Verify unauthenticated requests to feature_flags settings are rejected
- **Expected Status**: 401
- **Actual Status**: 401

<details>
<summary>curl command</summary>

```bash
curl -sk https://localhost:8000/api/gateway/v1/settings/feature_flags/
```
</details>

### 7. Token-based access confirms UI preference removed :white_check_mark: (PASS)
- **Category**: auth_variant
- **Description**: Verify the removal is consistent across auth methods (OAuth2 token vs session)
- **Expected Status**: 200
- **Actual Status**: 200

<details>
<summary>curl command</summary>

```bash
curl -sk -H 'Authorization: Bearer Nh3NmtyT1HZ1YhZ3Ey2MfNs7A1aQ5p' https://localhost:8000/api/gateway/v1/settings/feature_flags/
```
</details>

### 8. Unauthenticated access to feature flags list returns 401 :white_check_mark: (PASS)
- **Category**: negative
- **Description**: Verify unauthenticated requests to feature_flags endpoint are rejected
- **Expected Status**: 401
- **Actual Status**: 401

<details>
<summary>curl command</summary>

```bash
curl -sk https://localhost:8000/api/gateway/v1/feature_flags/
```
</details>


---

## Coverage Assessment

All 8 tests passed. The PR cleanly removes RUNTIME_FEATURE_FLAGS_UI from the preferences API without affecting sibling RUNTIME_FEATURE_FLAGS or the feature_flags endpoint. Tested across session auth, token auth, and unauthenticated access. No regressions found.

---


## Recommended Unit Tests

Edge cases discovered during DVT that should be covered by unit tests:

- **Test DB cleanup of orphaned preference row** — The PR removes the `register()` call but does not include a migration to delete the `RUNTIME_FEATURE_FLAGS_UI` row from the database. Existing deployments may retain stale data. DVT tests `#1` and `#7` confirm the API no longer exposes it, but a unit test should verify the DB row is cleaned up (or explicitly document that orphaned rows are acceptable).

---

## Methodology

This DVT was performed by an AI assistant against a live AAP Gateway instance
running locally via docker-compose. Tests were executed via curl against
`https://localhost:8000/` with self-signed TLS certificates.

Test categories:
- **happy_path**: Feature works as described in the PR
- **boundary**: Edge cases and limits
- **negative**: Invalid inputs, unauthorized access
- **regression**: Pre-existing functionality still works
- **auth**: Behavior across authentication variants (admin, user, token, unauthenticated)

File: aap_gateway_api/registered_preferences.py (hunk 1 above)
Comment (bhavenst) [reply aap_gateway_api/registered_preferences.py:520]: Yeah, nothing would trigger the removal.  Should be a custom removal migration, probably one-way since it was never used..

Comment (huffmanca) [reply aap_gateway_api/registered_preferences.py:520]: There's now a migration script to remove this row. Granted, I've never used a migration such as this (just adding/removing models, not rows within an existing model), so a double check of it would be appreciated.

Comment (bhavenst) [reply aap_gateway_api/registered_preferences.py:520]: Tested it, seems to work fine..

**Review by bhavenst — APPROVED** [review]

**Review by john-westcott-iv — APPROVED** [review]
## 📋 Re-Review: Approved

My previous comments have been addressed:

1. **Migration added** — `0021_remove_runtime_feature_flags_ui_preference.py` properly cleans up the orphaned DB row using `apps.get_model` with a filter on both `section` and `name`. The reverse migration includes a helpful warning. This was the blocking concern from my first review.

2. **Style nit** (parametrize tuple syntax) — pre-existing, not blocking.

**Files Reviewed**: 5 files (4 original + 1 new migration)
**Overall Assessment**: LGTM. Clean removal with proper DB cleanup.
